pkg/fileservice/memcache_stats_collector.go

```diff
@@ -0,0 +1,46 @@
+package fileservice
+
+import (
+	"github.com/matrixorigin/matrixone/pkg/util/metric"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+type memcacheStatsCollector struct {
+	memcache *MemCache
```

Comment:
Actually, do we really need a memcache_stats_collector?    Why not a system stats collector that once you register a metric (or a family of metrics), let system collector collects all families.

Reply:
1. Having one global registration of stats will violate this requirement: "init/register the stats counter only if the node needs it. For example, logservice does not need Fs or Memcache Stats."
2. Having one global registration of stats can cause import cycles.
3. Modules like `MemCache`& `DiskCache` already expose the Stats class individually. We are not polluting these Stats classes by adding Prometheus Collector Implementation details to these classes directly.
4. The current implementation is based on an existing Collector Logic defined by Prometheus for [Database](https://pkg.go.dev/github.com/prometheus/client_golang/prometheus/collectors#NewDBStatsCollector). This approach is clean because a) Stats Metric definitions (Counter, Gauge, etc) are defined in the callee side. b) We only register the Stats if the service needs it.

Reply:
Nah, one stats collector.   Need a central place to control all metrics.

Reply:
### Clarification
@fengttt I would like to clarify few things to ensure that, we both are on the same page, 
1. the `memcacheStatsCollector` is not creating a go-routine for collecting the MemCache Stats. It is only a `metric family` definition used by the Prometheus registry. **The name is a bit misleading. I will rename it more sensible.**
2. the registration of `metric family` to the global `dev_metric registry` happens via [this code](https://github.com/matrixorigin/matrixone/blob/68a3ebf06220b4aad7d6c9d5d8e381a92b34609a/pkg/fileservice/mem_cache.go#L43)
```go
mcStatsCollector := NewMemCacheStatsCollector(mc, "MemCache") //"MemCache" is the family name. mcStatsCollector is the family metric definition
metric.RegisterDevMetric(mcStatsCollector) //registers mcStats metric family definition to dev_registry
```
3. the central go-routine that collects all the registered `dev-metrics` every 15s is defined in [moDevMetricExporter](https://github.com/matrixorigin/matrixone/blob/68a3ebf06220b4aad7d6c9d5d8e381a92b34609a/pkg/util/metric/mometric/metric.go#L85) inside metrics pkg. We are reusing the existing code.
```go
moDevMetricExporter := newMetricExporter(metric.DefaultDevMetricRegistry, moDevMetricCollector, nodeUUID, role)
```
The `moDevMetricExporter` starts a single background go-routine which calls every the `metricFamily's` `gather()` method to fetch the stats at that point in time.

4. the collected metrics values are then written to log file using [moDevMetricCollector](https://github.com/matrixorigin/matrixone/blob/68a3ebf06220b4aad7d6c9d5d8e381a92b34609a/pkg/util/metric/mometric/metric_collector_log.go#L19) using a batch pipeline.

### My Thoughts
I still think that the `metric_family` definition should be close to the callee (in our case memcache and s3_fs) inorder to initialized the family correctly (without import cycles, with right mapping of CacheStats with metricFamily description) and avoid initialize metric_families that are not used by the service.

@aptend @xzxiong What are your thoughts on this?

Reply:
still, really need prometheus?   why 1 or 2 atomic int64 per metric is not good enough?

Reply:
Sure. I will remove the Prometheus dependency.